### PR TITLE
London | 26-ITP-May | Zadri Abdule | Sprint 3 | Implement & rewrite tests

### PR DESCRIPTION
**Pull request by Zadri415**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist
Added  Jest tests for Sprint-3 implement and rewrite (exercises 1–3)

### CODE REVIEW

File: Sprint-3/1-implement-and-rewrite-tests/implement/3-get-card-value.js [hunk 1]
```diff
@@ -21,8 +21,25 @@
...
+  if (!validSuits.includes(suit)) {
+    throw new Error("Invalid card");
+  }
+
+  const rank = card.slice(0, -1);
+  if (rank === "A") return 11;
+  if (rank === "J" || rank === "Q" || rank === "K") return 10;
+
+  const num = Number(rank);
+  if (Number.isInteger(num) && num >= 2 && num <= 10) return num;
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/implement/3-get-card-value.js]: In JavaScript, strings that represent valid numeric literals in the language can be safely converted to equivalent numbers. For examples, "0x02", "2.000", or "0002".

Does your function return the value you expected from each of the following function calls?

```
getCardValue("0x02♠");
getCardValue("2.000♠");
getCardValue("0002♠");
```

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js [hunk 2]
```diff
@@ -14,7 +14,33 @@ test(`should return "Acute angle" when (0 < angle < 90)`, () => {
...
+});
+
 // Case 6: Invalid angles
+test(`should return "Invalid angle" when invalid`, () => {
+  expect(getAngleType(-1)).toEqual("Invalid angle");
+  expect(getAngleType(0)).toEqual("Invalid angle");
+  expect(getAngleType("90")).toEqual("Invalid angle");
+  expect(getAngleType(1000)).toEqual("Invalid angle");
+  expect(getAngleType(9999.999)).toEqual("Invalid angle");
+});
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: - When a test fails with the message "... when invalid", it may be unclear what "invalid" actually refers to.
Can you revise the test description to make it more informative?

- Could also test the other boundary case, 360.

- Regarding the test `expect(getAngleType("90")).toEqual("Invalid angle")`:
    - It is not immediately clear why `"90"` is considered an invalid angle
    - Why `"90"` is considered invalid, but `"91"` is not?
    - It may imply the function is expected to check the type of the parameter

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js [hunk 3]
```diff
@@ -7,4 +7,36 @@ const isProperFraction = require("../implement/2-is-proper-fraction");
...
+});
+
+test(`should return true when numerator is zero and denominator is non-zero`, () => {
+  expect(isProperFraction(0, 1)).toEqual(true);
+  expect(isProperFraction(0, -1)).toEqual(true);
+  expect(isProperFraction(0, 100)).toEqual(true);
+  expect(isProperFraction(0, -100)).toEqual(true);
+});
+
+test(`should return true when absolute value of numerator is less than absolute value of denominator`, () => {
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js:21]: Note: Using pseudocode and notations like `abs(...)` or `| ... |` is a common and widely accepted practice for describing conditions in a concise manner.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js [hunk 4]
```diff
@@ -11,10 +11,34 @@ test(`Should return 11 when given an ace card`, () => {
 
 // Suggestion: Group the remaining test data into these categories:
 //   Number Cards (2-10)
+test(`Should return the numeric value for number cards`, () => {
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js:19]: Could also include a test to show that the function is expected to check the suit character.

**Review by cjyuan — COMMENTED** [review]
Very solid function implementation and tests.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js (hunk 4 above)
Comment (Zadri415) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js:19]: Thanks for the helpful feedback . I've updated the test description to make them clearer, added the missing boundary case for 360, revisit the "90" test so the invalid-input expectation makes more sense, and add a test to show that getCardValue checks the suit character too. I also appreciate the note about using mathematical notation in test descriptions.

File: Sprint-3/1-implement-and-rewrite-tests/implement/3-get-card-value.js [hunk 5]
```diff
@@ -21,8 +21,39 @@
 // After you have implemented the function, write tests to cover all the cases, and
 // execute the code to ensure all tests pass.
 
+const validSuits = ["♠", "♣", "♥", "♦"];
+const validRanks = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"];
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/implement/3-get-card-value.js:25]: What's the purpose of preparing `validRanks`? Could using it simplify the solution?

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js [hunk 6]
```diff
@@ -14,7 +14,37 @@ test(`should return "Acute angle" when (0 < angle < 90)`, () => {
...
+
 // Case 5: Reflex angles
+test(`should return "Reflex angle" when (180 < angle < 360)`, () => {
+  expect(getAngleType(181)).toEqual("Reflex angle");
+  expect(getAngleType(270)).toEqual("Reflex angle");
+  expect(getAngleType(359)).toEqual("Reflex angle");
+});
+
 // Case 6: Invalid angles
+test(`should return "Invalid angle" when angle is not greater than 0`, () => {
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: Why not simply express "not greater than 0" as `<= 0`? It's fine to use pseudocode or well-known notation in test descriptions when it makes them more concise. (Same for the "not less than 360")

In this case, `<=` is a widely recognized operator meaning "less than or equal to".

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js:1]: > ... add a test to show that getCardValue checks the suit character too. 

I could not find this test.

File: Sprint-3/1-implement-and-rewrite-tests/implement/3-get-card-value.js (hunk 5 above)
Comment (Zadri415) [reply Sprint-3/1-implement-and-rewrite-tests/implement/3-get-card-value.js:25]: To check if rank is valid, instead of comparing it to each value one by one.

Comment (cjyuan) [reply Sprint-3/1-implement-and-rewrite-tests/implement/3-get-card-value.js:25]: Sounds good. Why not implement what you described? Otherwise, it is best practice to remove all unused variables to keep the code clean.

**Comment by cjyuan** [comment]
Changes look good.

**Comment by Zadri415** [comment]
Thank you.
